File: example/basics1_display_robot_states.cpp
/**
 * @example basics1_display_robot_states.cpp
 * This tutorial does the very first thing: check connection with a pair of robots and print the
 * received robot states.
 * @copyright Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/drdk/robot_pair.hpp>
#include <flexiv/rdk/utility.hpp>
#include <spdlog/spdlog.h>

#include <iostream>
#include <thread>

using namespace flexiv;

/** @brief Print program usage help */
void PrintHelp()
{
    // clang-format off
    std::cout << "Required arguments: [left_robot_sn] [right_robot_sn]" << std::endl;
    std::cout << "    left_robot_sn: Serial number of the left robot to connect. Remove any space, e.g. Rizon4s-100001" << std::endl;
    std::cout << "    right_robot_sn: Serial number of the right robot to connect. Remove any space, e.g. Rizon4s-100002" << std::endl;
    std::cout << "Optional arguments: None" << std::endl;
    std::cout << std::endl;
    // clang-format on
}

/** @brief Print robot states data @ 1Hz */
void PrintRobotStates(drdk::RobotPair& robot_pair)
{
    while (true) {
        // Print all robot states in JSON format using the built-in ostream operator overloading
        spdlog::info("Left robot states:");
        std::cout << robot_pair.states().first << std::endl;
        spdlog::info("Right robot states:");
        std::cout << robot_pair.states().second << std::endl;

        // Print digital inputs
        spdlog::info("Left robot digital inputs:");
        std::cout << rdk::utility::Arr2Str(robot_pair.digital_inputs().first) << std::endl;
        spdlog::info("Right robot digital inputs:");
        std::cout << rdk::utility::Arr2Str(robot_pair.digital_inputs().second) << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

int main(int argc, char* argv[])
{
    // Program Setup
    // =============================================================================================
    // Parse parameters
    if (argc < 3 || rdk::utility::ProgramArgsExistAny(argc, argv, {"-h", "--help"})) {
        PrintHelp();
        return 1;
    }

    // Serial numbers of the left and right robot respectively
    std::string left_robot_sn = argv[1];
    std::string right_robot_sn = argv[2];

    // Print description
    spdlog::info(
        ">>> Tutorial description <<<\nThis tutorial does the very first thing: check connection "
        "with a pair of robots and print received robot states.\n");

    try {
        // RDK Initialization
        // =========================================================================================
        // Instantiate robot pair interface
        drdk::RobotPair robot_pair({left_robot_sn, right_robot_sn});

        // Clear fault on the connected robots if any
        if (robot_pair.fault()) {
            spdlog::warn("Fault occurred on one of the connected robots, trying to clear ...");
            // Try to clear the fault
            auto result = robot_pair.ClearFault();
            if (!result.first || !result.second) {
                spdlog::error("Fault cannot be cleared, exiting ...");
                return 1;
            }
            spdlog::info("Fault on the connected robot is cleared");
        }

        // Enable the pair of robots, make sure the E-stop is released before enabling
        spdlog::info("Enabling robots ...");
        robot_pair.Enable();

        // Wait for both robots to become operational
        while (!robot_pair.operational()) {
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        spdlog::info("Both robots are now operational");

        // Print States
        // =========================================================================================
        // Use std::thread to do scheduling so that this example can run on all OS, since not all OS
        // support rdk::Scheduler
        std::thread low_priority_thread(std::bind(PrintRobotStates, std::ref(robot_pair)));

        // Properly exit thread
        low_priority_thread.join();
    } catch (const std::exception& e) {
        spdlog::error(e.what());
        return 1;
    }

    return 0;
}

File: example_py/basics1_display_robot_states.py
#!/usr/bin/env python

"""basics1_display_robot_states.py

This tutorial does the very first thing: check connection with a pair of robots and print the
received robot states.
"""

__copyright__ = "Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved."
__author__ = "Flexiv"

import time
import argparse
import threading
import spdlog  # pip install spdlog
from flexivrdk import RobotStates  # pip install flexivrdk, for type definition
import flexivdrdk  # pip install flexivdrdk


def print_robot_states(robot_pair, logger, stop_event):
    """
    Print robot states data @ 1Hz.

    """

    while not stop_event.is_set():
        # Print all robot states, round all float values to 2 decimals
        left_robot_states, right_robot_states = robot_pair.states()
        logger.info("Left robot states:")
        # fmt: off
        print("{")
        print(f"q: {['%.2f' % i for i in left_robot_states.q]}",)
        print(f"theta: {['%.2f' % i for i in left_robot_states.theta]}")
        print(f"dq: {['%.2f' % i for i in left_robot_states.dq]}")
        print(f"dtheta: {['%.2f' % i for i in left_robot_states.dtheta]}")
        print(f"tau: {['%.2f' % i for i in left_robot_states.tau]}")
        print(f"tau_des: {['%.2f' % i for i in left_robot_states.tau_des]}")
        print(f"tau_dot: {['%.2f' % i for i in left_robot_states.tau_dot]}")
        print(f"tau_ext: {['%.2f' % i for i in left_robot_states.tau_ext]}")
        print(f"tcp_pose: {['%.2f' % i for i in left_robot_states.tcp_pose]}")
        print(f"tcp_velocity: {['%.2f' % i for i in left_robot_states.tcp_vel]}")
        print(f"flange_pose: {['%.2f' % i for i in left_robot_states.flange_pose]}")
        print(f"ft_sensor_raw: {['%.2f' % i for i in left_robot_states.ft_sensor_raw]}")
        print(f"ext_wrench_in_tcp: {['%.2f' % i for i in left_robot_states.ext_wrench_in_tcp]}")
        print(f"ext_wrench_in_world: {['%.2f' % i for i in left_robot_states.ext_wrench_in_world]}")
        print(f"ext_wrench_in_tcp_raw: {['%.2f' % i for i in left_robot_states.ext_wrench_in_tcp_raw]}")
        print(f"ext_wrench_in_world_raw: {['%.2f' % i for i in left_robot_states.ext_wrench_in_world_raw]}")
        print("}", flush= True)
        logger.info("Right robot states:")
        print("{")
        print(f"q: {['%.2f' % i for i in right_robot_states.q]}",)
        print(f"theta: {['%.2f' % i for i in right_robot_states.theta]}")
        print(f"dq: {['%.2f' % i for i in right_robot_states.dq]}")
        print(f"dtheta: {['%.2f' % i for i in right_robot_states.dtheta]}")
        print(f"tau: {['%.2f' % i for i in right_robot_states.tau]}")
        print(f"tau_des: {['%.2f' % i for i in right_robot_states.tau_des]}")
        print(f"tau_dot: {['%.2f' % i for i in right_robot_states.tau_dot]}")
        print(f"tau_ext: {['%.2f' % i for i in right_robot_states.tau_ext]}")
        print(f"tcp_pose: {['%.2f' % i for i in right_robot_states.tcp_pose]}")
        print(f"tcp_velocity: {['%.2f' % i for i in right_robot_states.tcp_vel]}")
        print(f"flange_pose: {['%.2f' % i for i in right_robot_states.flange_pose]}")
        print(f"ft_sensor_raw: {['%.2f' % i for i in right_robot_states.ft_sensor_raw]}")
        print(f"ext_wrench_in_tcp: {['%.2f' % i for i in right_robot_states.ext_wrench_in_tcp]}")
        print(f"ext_wrench_in_world: {['%.2f' % i for i in right_robot_states.ext_wrench_in_world]}")
        print(f"ext_wrench_in_tcp_raw: {['%.2f' % i for i in right_robot_states.ext_wrench_in_tcp_raw]}")
        print(f"ext_wrench_in_world_raw: {['%.2f' % i for i in right_robot_states.ext_wrench_in_world_raw]}")
        print("}", flush= True)
        # fmt: on

        # Print digital inputs
        left_din, right_din = robot_pair.digital_inputs()
        logger.info("Left robot digital inputs:")
        print(left_din)
        logger.info("Right robot digital inputs:")
        print(right_din)
        time.sleep(1)


def main():
    # Create an event to signal the thread to stop
    stop_event = threading.Event()

    # Program Setup
    # ==============================================================================================
    # Parse arguments
    argparser = argparse.ArgumentParser()
    argparser.add_argument(
        "left_robot_sn",
        help="Serial number of the left robot to connect. Remove any space, e.g. Rizon4s-100001",
    )
    argparser.add_argument(
        "right_robot_sn",
        help="Serial number of the right robot to connect. Remove any space, e.g. Rizon4s-100002",
    )
    args = argparser.parse_args()

    # Define alias
    logger = spdlog.ConsoleLogger("Example")

    # Print description
    logger.info(
        ">>> Tutorial description <<<\nThis tutorial does the very first thing: check connection "
        "with a pair of robots and print received robot states.\n"
    )

    try:
        # RDK Initialization
        # ==========================================================================================
        # Instantiate robot pair interface
        robot_pair = flexivdrdk.RobotPair([args.left_robot_sn, args.right_robot_sn])

        # Clear fault on the connected robot if any
        if robot_pair.fault():
            logger.warn("Fault occurred on the connected robot, trying to clear ...")
            # Try to clear the fault on both robots
            result = robot_pair.ClearFault()
            # If fault is not cleared on both robots
            if not (result[0] and result[1]):
                logger.error("Fault cannot be cleared, exiting ...")
                return 1
            logger.info("Fault on the connected robot is cleared")

        # Enable the pair of robots, make sure the E-stop is released before enabling
        logger.info("Enabling robots ...")
        robot_pair.Enable()

        # Wait for the robot to become operational
        while not robot_pair.operational():
            time.sleep(1)

        logger.info("Both robots are now operational")

    except Exception as e:
        # Print exception error message
        logger.error(str(e))

    # Print States
    # =============================================================================
    # Thread for printing robot states
    print_thread = threading.Thread(
        target=print_robot_states, args=[robot_pair, logger, stop_event]
    )
    print_thread.start()

    # Use main thread to catch keyboard interrupt and exit thread
    try:
        while not stop_event.is_set():
            time.sleep(0.1)
    except KeyboardInterrupt:
        # Send signal to exit thread
        logger.info("Stopping print thread")
        stop_event.set()

    # Wait for thread to exit
    print_thread.join()
    logger.info("Print thread exited")


if __name__ == "__main__":
    main()
